lib/go/test/flow_epoch_test.go

```diff
@@ -1516,3 +1517,203 @@ func TestEpochReset(t *testing.T) {
+		[]flow.Address{addresses[1]},
+		[]sdkcrypto.Signer{signers[1]},
+		false,
+	)
+
+	// Perform epoch recovery with a new epoch and epoch recover overwriting the current epoch.
+	t.Run("Can recover the epoch and have everything return to normal", func(t *testing.T) {
+		epochTimingConfigResult := executeScriptAndCheck(t, b, templates.GenerateGetEpochTimingConfigScript(env), nil)
+
+		var (
+			startView      uint64 = 100
+			stakingEndView uint64 = 120
+			endView        uint64 = 160
+			targetDuration uint64 = numEpochViews
+			targetEndTime  uint64 = expectedTargetEndTime(epochTimingConfigResult, startEpochCounter+1)
+		)
+
+		collectorClusters := make([]cadence.Value, 3)
+		collectorClusters[0] = cadence.NewArray([]cadence.Value{CadenceString("node_1"), CadenceString("node_2"), CadenceString("node_3")})
+		collectorClusters[1] = cadence.NewArray([]cadence.Value{CadenceString("node_4"), CadenceString("node_5"), CadenceString("node_6")})
+		collectorClusters[2] = cadence.NewArray([]cadence.Value{CadenceString("node_7"), CadenceString("node_8"), CadenceString("node_9")})
+
+		dkgPubKeys := []string{"pubkey_1"}
+		dkgPubKeysCdc := make([]cadence.Value, len(dkgPubKeys))
+		for i, key := range dkgPubKeys {
+			dkgPubKeysCdc[i], _ = cadence.NewString(key)
+		}
+
+		nodeIDs := make([]cadence.Value, len(ids))
+		for i, id := range ids {
+			nodeIDs[i], _ = cadence.NewString(id)
+		}
+
+		args := []cadence.Value{
+			cadence.NewUInt64(startView),
+			cadence.NewUInt64(stakingEndView),
+			cadence.NewUInt64(endView),
+			cadence.NewUInt64(targetDuration),
+			cadence.NewUInt64(targetEndTime),
+			cadence.NewArray(collectorClusters),        // collectorClusters
+			cadence.NewArray(make([]cadence.Value, 0)), // clusterQCVoteData
+			cadence.NewArray(dkgPubKeysCdc),
+			cadence.NewArray(nodeIDs),
+			cadence.NewBool(true), // recover EFM with a new epoch
+		}
+
+		tx = createTxWithTemplateAndAuthorizer(b, templates.GenerateRecoverEpochScript(env), idTableAddress)
+		for _, arg := range args {
+			tx.AddArgument(arg)
+		}
+
+		signAndSubmit(
+			t, b, tx,
+			[]flow.Address{idTableAddress},
+			[]sdkcrypto.Signer{IDTableSigner},
+			false,
+		)
+
+		advanceView(t, b, env, idTableAddress, IDTableSigner, 1, "BLOCK", false)
+
+		// seed is not manually set when recovering the epoch, it is randomly generated
+		metadataFields := getEpochMetadata(t, b, env, cadence.NewUInt64(startEpochCounter+1))
+		seed := strings.ReplaceAll(metadataFields["seed"].String(), `"`, "")
+		expectedMetadata := EpochMetadata{
+			counter:               startEpochCounter + 1,
+			seed:                  seed,
+			startView:             startView,
+			endView:               endView,
+			stakingEndView:        stakingEndView,
+			totalRewards:          "0.0",
+			rewardsBreakdownArray: 0,
+			rewardsPaid:           false,
+			collectorClusters:     nil,
+			clusterQCs:            nil,
+			dkgKeys:               dkgPubKeys,
+		}
+		verifyEpochMetadata(t, b, env, expectedMetadata)
```

Comment:
Can we also validate that `curEpochCounter = startEpochCounter+1`

Reply:
https://github.com/onflow/flow-core-contracts/pull/420/commits/614f489fb0448a5278b4eec15cee7742701370ae